### PR TITLE
Mystery Gift script commands

### PR DESCRIPTION
**Pull request by PikalaxALT**
Must come after #271

### CODE REVIEW

File: src/scrcmd_11.c [hunk 1]
```diff
@@ -177,16 +183,131 @@ BOOL ScrCmd_MysteryGift(ScriptContext* ctx) {
...
+
+    if (mgData->fixedOT == OT_ID_PLAYER_ID) {
+        String* playerName = PlayerProfile_GetPlayerName_NewString(profile, HEAP_ID_32);
+        u32 trainerId = PlayerProfile_GetTrainerID(profile);
+        BOOL gender = PlayerProfile_GetTrainerGender(profile);
+
+        tmpPokemon = AllocMonZeroed(HEAP_ID_32);
+        #ifdef UBFIX
+        GF_ASSERT(pokemon != NULL);
+        #endif
```
Comment (tgsm) [inline src/scrcmd_11.c]:
```suggestion
#ifdef UBFIX
        GF_ASSERT(pokemon != NULL);
#endif
```

File: src/scrcmd_mystery_gift.c [hunk 2]
```diff
@@ -311,3 +318,372 @@ void MGGive_Mon(FieldSystem* fieldSys, MysteryGiftData* unused) {
...
+void MGGive_PokewalkerCourse(FieldSystem* fieldSys, MysteryGiftData* unused) {
+    (void)unused;
+    POKEWALKER* pokeWalker = Save_Pokewalker_Get(fieldSys->saveData);
+    u8 courseNo = FieldSystem_GetDataOfNextMG(fieldSys)->pokewalkerCourse;
+    if (courseNo < WALKER_COURSE_MAX) {
+        Pokewalker_UnlockCourse(pokeWalker, courseNo);
+    }
+}
+
+void MGMesageSuccess_PokewalkerCourse(struct GetMysteryGiftGmmState* gmmState, u16* pMsgBank, u16* pMsgNum) {
```
Comment (tgsm) [inline src/scrcmd_mystery_gift.c]:
```suggestion
void MGMessageSuccess_PokewalkerCourse(struct GetMysteryGiftGmmState* gmmState, u16* pMsgBank, u16* pMsgNum) {
```

File: src/scrcmd_mystery_gift.c [hunk 3]
```diff
@@ -311,3 +318,372 @@ void MGGive_Mon(FieldSystem* fieldSys, MysteryGiftData* unused) {
...
+    u8* mgData = &FieldSystem_GetDataOfNextMG(gmmState->fieldSys)->pokewalkerCourse;
+    *pMsgBank = NARC_msg_msg_0209_bin;
+    *pMsgNum = msg_0209_00019;
+    BufferPlayersName(gmmState->msgFormat, 0, Save_PlayerData_GetProfileAddr(gmmState->fieldSys->saveData));
+    BufferPokewalkerCourseName(gmmState->msgFormat, 1, *mgData);
+}
+
+void MGMesageFailure_PokewalkerCourse(struct GetMysteryGiftGmmState* gmmState, u16* pMsgBank, u16* pMsgNum) {
```
Comment (tgsm) [inline src/scrcmd_mystery_gift.c]: ditto

**Review by tgsm — CHANGES_REQUESTED** [review]

File: include/fashion_case.h [hunk 4]
```diff
@@ -27,17 +27,17 @@ typedef struct SaveFashionData { //todo: rename to SaveFashionData
...
 BOOL sub_0202BA08(SaveFashionData *fashionData, int a1);
 SaveFashionDataSub *sub_0202B9B8(SaveFashionData *fashionData, int a1);
 void sub_0202BD7C(SaveFashionDataSub *fashionDataSub, int a1);
-void sub_0202BB08(FashionCase *a0, int a1, int a2);
-int sub_0202BA2C(FashionCase *a0, int a1, int a2);
-u32 sub_0202BA70(FashionCase *a0, int a1);
-void sub_0202BBD8(FashionCase *a0, int a1);
-int sub_0202BA5C(FashionCase *a0, int a1);
-void sub_0202BB7C(FashionCase *a0, int a1, int a2);
+void GiveFashionItem(FashionCase *fashionCase, int id, int quantity);
```
Comment (adrienntindall) [inline include/fashion_case.h:36]: FashionCase_GiveFashionItem

File: include/fashion_case.h [hunk 5]
```diff
@@ -27,17 +27,17 @@ typedef struct SaveFashionData { //todo: rename to SaveFashionData
...
+int sub_0202BA2C(FashionCase *fashionCase, int a1, int a2);
+u32 sub_0202BA70(FashionCase *fashionCase, int a1);
+void GiveContestBackground(FashionCase *fashionCase, int id);
```
Comment (adrienntindall) [inline include/fashion_case.h:39]: FashionCase_GiveContestBackground

**Review by adrienntindall — CHANGES_REQUESTED** [review]
Generally gonna revisit this once the previous branch is merged since I don't want to repeat myself

File: src/scrcmd_mystery_gift.c (hunk 2 above)
Comment (PikalaxALT) [reply src/scrcmd_mystery_gift.c]: unown_questionmark_front.png

**Review by tgsm — DISMISSED** [review]

File: src/trainer_memo.c [hunk 6]
```diff
@@ -695,80 +698,80 @@ void BoxMonSetTrainerMemo(BoxPokemon *boxMon, PlayerProfile *profile, int strat,
...
-        sub_0208F550(boxMon, mapsec, FALSE);
-        sub_0208F5C8(boxMon, TRUE);
+        SetBoxMonMetDateAndLocation(boxMon, mapsec, SETMETDATEPARAM_EGG);
+        ClearBoxMonMetDateAndLocation(boxMon, SETMETDATEPARAM_MON);
         break;
     case 1:
         if (!GetBoxMonData(boxMon, MON_DATA_IS_EGG, NULL)) {
-            sub_0208F5C8(boxMon, FALSE);
-            sub_0208F550(boxMon, sub_02017FE4(MAPSECTYPE_GIFT, 1), TRUE);
+            ClearBoxMonMetDateAndLocation(boxMon, SETMETDATEPARAM_EGG);
```
Comment (adrienntindall) [inline src/trainer_memo.c:729]: BoxMon_ClearMetDateAndLocation

File: src/trainer_memo.c [hunk 7]
```diff
@@ -695,80 +698,80 @@ void BoxMonSetTrainerMemo(BoxPokemon *boxMon, PlayerProfile *profile, int strat,
...
+            SetBoxMonMetDateAndLocation(boxMon, sub_02017FE4(MAPSECTYPE_GIFT, MAPLOC(METLOC_LINK_TRADE)), SETMETDATEPARAM_MON);
```
Comment (adrienntindall) [inline src/trainer_memo.c:730]: BoxMon_SetMetDateAndLocation

**Review by adrienntindall — CHANGES_REQUESTED** [review]

File: include/unk_020290B4.h [hunk 8]
```diff
@@ -2,9 +2,12 @@
 #define POKEHEARTGOLD_UNK_020290B4_H
 
 #include "save.h"
+#include "pm_string.h"
 
 BOOL sub_020291A4(SaveData *saveData, int a1);
 u32 sub_02029264(void *a0);
 BOOL sub_020290FC(void*, int);
+void sub_020291D4(SaveData* saveData, u16* a1);
```
Comment (red031000) [inline include/unk_020290B4.h:10]: pointer alignment plssss

File: main.lsf [hunk 9]
```diff
@@ -173,7 +173,7 @@ Static main
 	Object src/unk_0204A3F4.o
 	Object src/unk_0204B538.o
 	Object src/scrcmd_apricorn.o
-	Object asm/scrcmd_11.o
+	Object src/field/scrcmd_mystery_gift.o
```
Comment (red031000) [inline main.lsf:176]: THIS IS NOT IN FIELD

File: src/field/scrcmd_mystery_gift.c [hunk 10]
```diff
@@ -0,0 +1,684 @@
...
+#include "fashion_case.h"
+#include "unk_020290B4.h"
+#include "pokewalker.h"
+#include "trainer_memo.h"
+#include "constants/map_sections.h"
+#include "msgdata/msg.naix"
+#include "msgdata/msg/msg_0209.h"
+
+struct GetMysteryGiftGmmState {
+    FieldSystem* fieldSys;
```
Comment (red031000) [inline src/field/scrcmd_mystery_gift.c:20]: use fieldSystem instead of fieldSys

File: src/field/scrcmd_mystery_gift.c [hunk 11]
```diff
@@ -0,0 +1,684 @@
...
+    MessageFormat* msgFormat;
+    MysteryGiftData* mgData;
+};
+
+struct ScriptMysteryGiftFuncs {
+    BOOL (*check)(FieldSystem*, MysteryGiftData*);
```
Comment (red031000) [inline src/field/scrcmd_mystery_gift.c:26]: pointer alignment

File: src/field/scrcmd_mystery_gift.c [hunk 12]
```diff
@@ -0,0 +1,684 @@
...
+    void (*give)(FieldSystem*, MysteryGiftData*);
+    void (*messageSuccess)(struct GetMysteryGiftGmmState*, u16*, u16*);
+    void (*messageFailure)(struct GetMysteryGiftGmmState*, u16*, u16*);
+};
+
+void FieldSystem_InitGetMysteryGiftGmmState(struct GetMysteryGiftGmmState* state, FieldSystem* fieldSys, MessageFormat* msgFormat, MysteryGiftData* mgData);
```
Comment (red031000) [inline src/field/scrcmd_mystery_gift.c:32]: if these are static they need the static keyword, if they're extern they need the extern keyword

File: src/field/scrcmd_mystery_gift.c [hunk 13]
```diff
@@ -0,0 +1,684 @@
...
+#include "sys_vars.h"
+#include "fashion_case.h"
+#include "unk_020290B4.h"
+#include "pokewalker.h"
+#include "trainer_memo.h"
+#include "constants/map_sections.h"
+#include "msgdata/msg.naix"
+#include "msgdata/msg/msg_0209.h"
+
+struct GetMysteryGiftGmmState {
```
Comment (red031000) [inline src/field/scrcmd_mystery_gift.c:19]: typedef these structs

File: src/field/scrcmd_mystery_gift.c (hunk 12 above)
Comment (red031000) [reply src/field/scrcmd_mystery_gift.c:32]: also do not use struct keyword for struct params

File: src/field/scrcmd_mystery_gift.c [hunk 14]
```diff
@@ -0,0 +1,684 @@
...
+void FieldSystem_SetQueuedMGReceived(FieldSystem* fieldSys);
+BOOL MGCheck_PartySpace(FieldSystem* fieldSys, MysteryGiftData* mgData);
+void MGGive_Mon(FieldSystem* fieldSys, MysteryGiftData* mgData);
+void MGMessageSuccess_GiveMon(struct GetMysteryGiftGmmState* fsysUnk80, u16* pMsgBank, u16* pMsgNum);
+void MGMessageFailure_GiveMon(struct GetMysteryGiftGmmState* fsysUnk80, u16* pMsgBank, u16* pMsgNum);
+void MGGive_Egg(FieldSystem* fieldSys, MysteryGiftData* mgData);
+void MGMessageSuccess_Egg(struct GetMysteryGiftGmmState* fsysUnk80, u16* pMsgBank, u16* pMsgNum);
+BOOL MGCheck_Item(FieldSystem* fieldSys, MysteryGiftData* mgData);
+void MGGive_Item(FieldSystem* fieldSys, MysteryGiftData* mgData);
+void MGMessageSuccess_Item(struct GetMysteryGiftGmmState* fsysUnk80, u16* pMsgBank, u16* pMsgNum);
```
Comment (red031000) [inline src/field/scrcmd_mystery_gift.c:44]: document fsysUnk80 as well

File: src/field/scrcmd_mystery_gift.c [hunk 15]
```diff
@@ -0,0 +1,684 @@
...
+};
+
+void FieldSystem_InitGetMysteryGiftGmmState(struct GetMysteryGiftGmmState* state, FieldSystem* fieldSys, MessageFormat* msgFormat, MysteryGiftData* mgData) {
+    state->fieldSys = fieldSys;
+    state->msgFormat = msgFormat;
+    state->mgData = mgData;
+}
+
+int FieldSystem_GetTagOfNextMG(FieldSystem* fieldSys) {
+    (void)fieldSys;
```
Comment (red031000) [inline src/field/scrcmd_mystery_gift.c:114]: not needed

File: src/field/scrcmd_mystery_gift.c [hunk 16]
```diff
@@ -0,0 +1,684 @@
...
+        break;
+    case SCR_MG_END:
+        SaveMGDataPtr_End(ctx->fieldSystem->saveData, FALSE);
+        break;
+    case SCR_MG_END2:
+        SaveMGDataPtr_End(ctx->fieldSystem->saveData, TRUE);
+        break;
+    case SCR_MG_HAS_GIFT: {
+        u16* ptr = ScriptGetVarPointer(ctx);
+        if (FieldSystem_GetTagOfNextMG(ctx->fieldSystem) != MG_TAG_INVALID) {
```
Comment (red031000) [inline src/field/scrcmd_mystery_gift.c:141]: does `*ptr = FieldSystem_GetTagOfNextMG(ctx->fieldSystem) != MG_TAG_INVALID` work?

File: src/field/scrcmd_mystery_gift.c [hunk 17]
```diff
@@ -0,0 +1,684 @@
...
+        break;
+    }
+    case SCR_MG_GET_TYPE: {
+        u16* ptr = ScriptGetVarPointer(ctx);
+        *ptr = FieldSystem_GetTagOfNextMG(ctx->fieldSystem);
+        break;
+    }
+    case SCR_MG_CAN_RECEIVE: {
+        u16* ptr = ScriptGetVarPointer(ctx);
+        const struct ScriptMysteryGiftFuncs* pFunc = &sScriptMysteryGiftActionTable[FieldSystem_GetTagOfNextMG(ctx->fieldSystem) - 1];
```
Comment (red031000) [inline src/field/scrcmd_mystery_gift.c:155]: no struct

File: src/field/scrcmd_mystery_gift.c [hunk 18]
```diff
@@ -0,0 +1,684 @@
...
+        struct GetMysteryGiftGmmState gmmState;
+        const struct ScriptMysteryGiftFuncs* pFunc = &sScriptMysteryGiftActionTable[FieldSystem_GetTagOfNextMG(ctx->fieldSystem) - 1];
+        MessageFormat** pMsgFormat = FieldSysGetAttrAddr(ctx->fieldSystem, SCRIPTENV_MESSAGE_FORMAT);
+        u16* pMsgBank = ScriptGetVarPointer(ctx);
+        u16* pMsgNum = ScriptGetVarPointer(ctx);
+        FieldSystem_InitGetMysteryGiftGmmState(&gmmState, ctx->fieldSystem, *pMsgFormat, FieldSystem_GetDataOfNextMG(ctx->fieldSystem));
+        pFunc->messageFailure(&gmmState, pMsgBank, pMsgNum);
+        break;
+    }
+    }
```
Comment (red031000) [inline src/field/scrcmd_mystery_gift.c:184]: two brackets on the same indentation level across two lines physically *hurts*

File: src/field/scrcmd_mystery_gift.c [hunk 19]
```diff
@@ -0,0 +1,684 @@
...
+    return FALSE;
+}
+
+BOOL MGCheck_PartySpace(FieldSystem* fieldSys, MysteryGiftData* unused) {
```
Comment (red031000) [inline src/field/scrcmd_mystery_gift.c:188]: part of me thinks these should be named MysteryGift_Check_PartySpace

Comment (red031000) [reply src/field/scrcmd_mystery_gift.c:188]: this entire file is confused, it's partly scrcmd, partly mystery gift

File: src/field/scrcmd_mystery_gift.c [hunk 20]
```diff
@@ -0,0 +1,684 @@
...
+
+    gender = GetMonData(pokemon, MON_DATA_HP_IV, NULL)
+        + GetMonData(pokemon, MON_DATA_ATK_IV, NULL)
+        + GetMonData(pokemon, MON_DATA_DEF_IV, NULL)
+        + GetMonData(pokemon, MON_DATA_SPEED_IV, NULL)
+        + GetMonData(pokemon, MON_DATA_SPATK_IV, NULL)
+        + GetMonData(pokemon, MON_DATA_SPDEF_IV, NULL);
+    if (gender == 0) {
+        u16 ivRand = LCRandom();
+        rand = ivRand & 0x1Fu;
```
Comment (red031000) [inline src/field/scrcmd_mystery_gift.c:240]: any significance to these values? or are they just arbritrary

File: src/field/scrcmd_mystery_gift.c [hunk 21]
```diff
@@ -0,0 +1,684 @@
...
+    }
+
+    if (mgData->fixedOT == OT_ID_PLAYER_ID) {
+        String* playerName = PlayerProfile_GetPlayerName_NewString(profile, HEAP_ID_32);
+        u32 trainerId = PlayerProfile_GetTrainerID(profile);
+        BOOL gender = PlayerProfile_GetTrainerGender(profile);
+
+        tmpPokemon = AllocMonZeroed(HEAP_ID_32);
+#ifdef UBFIX
+        GF_ASSERT(pokemon != NULL);
```
Comment (red031000) [inline src/field/scrcmd_mystery_gift.c:294]: why would this fix UB? it'd just error if it was null, and not actually fix the issue

File: src/field/scrcmd_mystery_gift.c [hunk 22]
```diff
@@ -0,0 +1,684 @@
...
+        SetMonData(tmpPokemon, MON_DATA_OT_NAME_2, playerName);
+        SetMonData(tmpPokemon, MON_DATA_OTID, &trainerId);
+        SetMonData(tmpPokemon, MON_DATA_MET_GENDER, &gender);
+        pokemon = tmpPokemon;
+        String_Delete(playerName);
+    }
+
+    MonSetTrainerMemo(pokemon, profile, 4, sub_02017FE4(MAPSECTYPE_EXTERNAL, eggMetLocation), HEAP_ID_32);
+    if (GetMonData(pokemon, MON_DATA_SPECIES, NULL) == SPECIES_ARCEUS && GetMonData(pokemon, MON_DATA_FATEFUL_ENCOUNTER, NULL) == TRUE && !Save_VarsFlags_GetVar404C(vars_flags)) {
+        Save_VarsFlags_SetVar404C(vars_flags, TRUE);
```
Comment (red031000) [inline src/field/scrcmd_mystery_gift.c:306]: enough info to document the var?

File: src/field/scrcmd_mystery_gift.c [hunk 23]
```diff
@@ -0,0 +1,684 @@
...
+    *pMsgBank = NARC_msg_msg_0209_bin;
+    *pMsgNum = msg_0209_00031;
+}
+
+BOOL MGCheck_MonDeco(FieldSystem* fieldSys, MysteryGiftData* unused) {
+    (void)unused;
+    MysteryGiftMonDecorationTag* mgData = &FieldSystem_GetDataOfNextMG(fieldSys)->monDecoration;
+    int id = mgData->id;
+    switch (mgData->kind) {
+    case 1:
```
Comment (red031000) [inline src/field/scrcmd_mystery_gift.c:434]: constants?

**Review by red031000 — CHANGES_REQUESTED** [review]
*please please **please*** abide by the style
there are things I disagree with in the style guide as well, but I still abide by it
the code would be an absolute mess if everyone just did what they wanted to do

File: src/field/scrcmd_mystery_gift.c (hunk 13 above)
Comment (PikalaxALT) [reply src/field/scrcmd_mystery_gift.c:19]: not required

File: src/field/scrcmd_mystery_gift.c (hunk 10 above)
Comment (PikalaxALT) [reply src/field/scrcmd_mystery_gift.c:20]: unnecessarily verbose

File: src/field/scrcmd_mystery_gift.c (hunk 11 above)
Comment (PikalaxALT) [reply src/field/scrcmd_mystery_gift.c:26]: strong disagree

File: src/field/scrcmd_mystery_gift.c (hunk 12 above)
Comment (PikalaxALT) [reply src/field/scrcmd_mystery_gift.c:32]: disagree with the second comment

File: src/field/scrcmd_mystery_gift.c (hunk 10 above)
Comment (red031000) [reply src/field/scrcmd_mystery_gift.c:20]: please be consistent with existing code

File: src/field/scrcmd_mystery_gift.c (hunk 11 above)
Comment (red031000) [reply src/field/scrcmd_mystery_gift.c:26]: style disagrees with your disagreement

File: src/field/scrcmd_mystery_gift.c (hunk 12 above)
Comment (red031000) [reply src/field/scrcmd_mystery_gift.c:32]: style again, this was majority agreed

File: src/field/scrcmd_mystery_gift.c (hunk 18 above)
Comment (PikalaxALT) [reply src/field/scrcmd_mystery_gift.c:184]: that's a you problem

File: src/field/scrcmd_mystery_gift.c (hunk 17 above)
Comment (PikalaxALT) [reply src/field/scrcmd_mystery_gift.c:155]: ?????????????????

File: src/field/scrcmd_mystery_gift.c (hunk 19 above)
Comment (PikalaxALT) [reply src/field/scrcmd_mystery_gift.c:188]: i do not share that part of you

File: src/field/scrcmd_mystery_gift.c (hunk 16 above)
Comment (PikalaxALT) [reply src/field/scrcmd_mystery_gift.c:141]: no. no it does not.

File: src/field/scrcmd_mystery_gift.c (hunk 20 above)
Comment (PikalaxALT) [reply src/field/scrcmd_mystery_gift.c:240]: not sure what you're asking. did this comment end up on the wrong line?

File: src/field/scrcmd_mystery_gift.c (hunk 21 above)
Comment (PikalaxALT) [reply src/field/scrcmd_mystery_gift.c:294]: if there's a memory allocation error, nothing you do will let the game continue to run in a valid state. the best course of action is to error out.

File: src/field/scrcmd_mystery_gift.c (hunk 22 above)
Comment (PikalaxALT) [reply src/field/scrcmd_mystery_gift.c:306]: not sure, possibly out of scope still

**Review by red031000 — CHANGES_REQUESTED** [review]
to merge: apply clang format to changes in unk_020290B4.h, then move scrcmd_mystery_gift.c out of the field folder

**Review by red031000 — APPROVED** [review]

**Review by adrienntindall — APPROVED** [review]

**Pull request merged**